aws/rust-runtime/aws-config/src/environment/timeout_config.rs

```diff
@@ -0,0 +1,231 @@
+use aws_types::os_shim_internal::Env;
+use std::time::Duration;
+
+const ENV_VAR_CONNECT_TIMEOUT: &str = "AWS_CONNECT_TIMEOUT";
+const ENV_VAR_TLS_NEGOTIATION_TIMEOUT: &str = "AWS_TLS_NEGOTIATION_TIMEOUT";
+const ENV_VAR_READ_TIMEOUT: &str = "AWS_READ_TIMEOUT";
+const ENV_VAR_API_CALL_ATTEMPT_TIMEOUT: &str = "AWS_API_CALL_ATTEMPT_TIMEOUT";
+const ENV_VAR_API_CALL_TIMEOUT: &str = "AWS_API_CALL_TIMEOUT";
+
+/// Load a timeout_config from environment variables
+///
+/// This provider will check the values of the following variables in order to build a `TimeoutConfig`
+///
+/// - AWS_CONNECT_TIMEOUT
+/// - AWS_TLS_NEGOTIATION_TIMEOUT
+/// - AWS_READ_TIMEOUT
+/// - AWS_API_CALL_ATTEMPT_TIMEOUT
+/// - AWS_API_CALL_TIMEOUT
+///
+/// If at least one of these is set to a valid value, construction will succeed.
+#[derive(Debug, Default)]
+pub struct EnvironmentVariableTimeoutConfigProvider {
+    env: Env,
+}
+
+impl EnvironmentVariableTimeoutConfigProvider {
+    /// Create a new `EnvironmentVariableTimeoutConfigProvider`
+    pub fn new() -> Self {
+        EnvironmentVariableTimeoutConfigProvider { env: Env::real() }
+    }
+
+    #[doc(hidden)]
+    /// Create an timeout_config provider from a given `Env`
+    ///
+    /// This method is used for tests that need to override environment variables.
+    pub fn new_with_env(env: Env) -> Self {
+        EnvironmentVariableTimeoutConfigProvider { env }
+    }
+
+    /// Attempt to create a new `TimeoutConfig` from environment variables
+    pub fn timeout_config_builder(&self) -> Result<TimeoutConfigBuilder, TimeoutConfigError> {
+        let connect_timeout = construct_timeout_from_env_var(&self.env, ENV_VAR_CONNECT_TIMEOUT)?
+            .map(Duration::from_secs_f32);
+        let tls_negotiation_timeout =
+            construct_timeout_from_env_var(&self.env, ENV_VAR_TLS_NEGOTIATION_TIMEOUT)?
+                .map(Duration::from_secs_f32);
+        let read_timeout = construct_timeout_from_env_var(&self.env, ENV_VAR_READ_TIMEOUT)?
+            .map(Duration::from_secs_f32);
+        let api_call_attempt_timeout =
+            construct_timeout_from_env_var(&self.env, ENV_VAR_API_CALL_ATTEMPT_TIMEOUT)?
+                .map(Duration::from_secs_f32);
+        let api_call_timeout = construct_timeout_from_env_var(&self.env, ENV_VAR_API_CALL_TIMEOUT)?
+            .map(Duration::from_secs_f32);
+
+        let mut builder = TimeoutConfigBuilder::new();
+        builder
+            .set_connect_timeout(connect_timeout)
+            .set_tls_negotiation_timeout(tls_negotiation_timeout)
+            .set_read_timeout(read_timeout)
+            .set_api_call_attempt_timeout(api_call_attempt_timeout)
+            .set_api_call_timeout(api_call_timeout);
+
+        Ok(builder)
+    }
+}
+
+const SET_BY: &str = "environment variable";
+
+fn construct_timeout_from_env_var(env: &Env, var: &str) -> Result<Option<f32>, TimeoutConfigError> {
+    // TODO do I really need to clone this?
+    let var = var.to_owned();
```

Comment:
you probably don't need a clone but you might need something other than `into()` below